python/pyspark/sql/pandas/group_ops.py

```diff
@@ -803,27 +806,30 @@ def applyInArrow(
         Applies a function to each cogroup using Arrow and returns the result
         as a `DataFrame`.
 
-        The function should take two `pyarrow.Table`\\s and return another
-        `pyarrow.Table`. Alternatively, the user can pass a function that takes
-        a tuple of `pyarrow.Scalar` grouping key(s) and the two `pyarrow.Table`\\s.
-        For each side of the cogroup, all columns are passed together as a
-        `pyarrow.Table` to the user-function and the returned `pyarrow.Table` are combined as
-        a :class:`DataFrame`.
+        The function can take one of two forms: It can take two `pyarrow.Table`\\s and return a
+        `pyarrow.Table`, or it can take two iterators of `pyarrow.RecordBatch` and yield
+        `pyarrow.RecordBatch`. Alternatively, each form can take a tuple of `pyarrow.Scalar`
+        as the first argument in addition to the input type above. For each cogroup, all columns
+        are passed together in the `pyarrow.Table` or `pyarrow.RecordBatch`, and the returned
+        `pyarrow.Table` or iterator of `pyarrow.RecordBatch` are combined as a :class:`DataFrame`.
 
         The `schema` should be a :class:`StructType` describing the schema of the returned
```

Comment:
We should add an implementation note here saying that

    The content of the first iterator of `pyarrow.RecordBatch` will be hold entirely in memory,
    while for the second iterator only the current `pyarrow.RecordBatch` is kept in memory.
    Users should use the dataframe that they anticipate to produce larger groups as the
    right argument of the cogroup to reduce memory footprint.

That way, users can already benefit from the iterator interface. Loading only one of both groups into memory is an improvement over the table interface.

Reply:
Maybe a follow-up PR to keep this one tidy?

Reply:
Yeah I can look into that in my follow on that actually does the batching on the JVM input side (unless others just want me to merge that into this branch)

Reply:
@zhengruifeng this is a potential way to address your comment above. Just load the left side fully into memory, but iteratively load the right side, so you at least don't need both in memory at once.